jax/experimental/pjit.py

```diff
@@ -88,6 +89,10 @@ def wrapped(*args, **kwargs):
         lambda: tuple(flatten_axes("pjit out_axis_resources", out_tree(),
                                    out_axis_resources)),
         closure=out_axis_resources)
+    _check_shapes_against_resources("pjit arguments", resource_env,
+                                    args_flat, in_axis_resources_flat)
+    flat_fun = _check_output_shapes(flat_fun, resource_env,
+                                    out_axis_resources_thunk)
 
     out = pjit_call_p.bind(
```

Comment:
Can you call `_check_output_shapes` after this bind call instead of using a transform? Maybe that's simpler?

Reply:
Sadly we can't. We want to check the output shapes at the very moment the user function returns, before anything inside `mesh_callable` starts tiling/untiling and asserting on divisibility constraints. So it feels safer to me to insert is through this transform.